### PR TITLE
Servlet Link Access Counter

### PR DESCRIPTION
**Pull request by Stanislava88**
In this PR i have solved the first task: created servlet that count number link is visited 
linkaccesscounter

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.12</version>
+      <scope>test</scope>
+    </dependency>
+
+
+    <!--jMock-->
```
Comment (freekman) [inline pom.xml]: just to know we usually separate the dependencies in groups
Ex: for the tests
```js
 +    <!--Test-->
<dependency>1
<dependency>2
.....
````

File: src/com/clouway/adapter/server/jetty/Jetty.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.clouway.links.ServletAccessCounter;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+
+import javax.servlet.ServletContext;
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+
+
```
Comment (freekman) [inline src/com/clouway/adapter/server/jetty/Jetty.java]: .

File: src/com/clouway/links/ServletAccessCounter.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import java.io.PrintWriter;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * @author Stanislava Kaukova(sisiivanovva@gmail.com)
+ */
+public class ServletAccessCounter extends HttpServlet {
+  private Map<String, Integer> links = new ConcurrentHashMap<String, Integer>();
+
```
Comment (freekman) [inline src/com/clouway/links/ServletAccessCounter.java]: .

File: src/com/clouway/links/ServletAccessCounter.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  @Override
+  public void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+    String link = req.getParameter("link");
+    countAccess(link);
```
Comment (freekman) [inline src/com/clouway/links/ServletAccessCounter.java]: maybe ```countLinkAccess``` will be beter here

File: src/com/clouway/links/ServletAccessCounter.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  public String getHtml(String path, PrintWriter writer) {
+    StringBuffer buffer = new StringBuffer();
+
+    try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(path)))) {
+      String line;
+      while ((line = reader.readLine()) != null) {
+        buffer.append(line);
+      }
+
+      HtmlTemplate template = new HtmlTemplate(buffer.toString());
```
Comment (freekman) [inline src/com/clouway/links/ServletAccessCounter.java]: if you assign ```buffer.toString()``` to a variable ``` String html``` would be better

File: src/com/clouway/links/ServletAccessCounter.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      replace(template);
```
Comment (freekman) [inline src/com/clouway/links/ServletAccessCounter.java]: there is no need of the func replace move the logic here

File: src/com/clouway/links/ServletAccessCounter.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+      String evaluationValue = template.evaluate();
+      buffer.append(evaluationValue);
+
+      writer.write(evaluationValue);
+      writer.flush();
+
+      return buffer.toString();
+    } catch (IOException e) {
+      return null;
```
Comment (freekman) [inline src/com/clouway/links/ServletAccessCounter.java]: why don't you return empty string insted

File: tests/com/clouway/templates/ServletCounterTest.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ */
+public class ServletCounterTest {
+  public JUnitRuleMockery context = new JUnitRuleMockery();
+
+  private HttpServletRequest request = context.mock(HttpServletRequest.class);
+  private HttpServletResponse response = context.mock(HttpServletResponse.class);
+  private PrintWriter writer;
+  private ByteArrayOutputStream outputStream;
+
+  public ServletCounterTest() throws IOException {
```
Comment (freekman) [inline tests/com/clouway/templates/ServletCounterTest.java]: .

File: tests/com/clouway/templates/ServletCounterTest.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.hamcrest.core.StringContains.containsString;
+
+/**
+ * @author Stanislava Kaukova(sisiivanovva@gmail.com)
+ */
+public class ServletCounterTest {
+  public JUnitRuleMockery context = new JUnitRuleMockery();
```
Comment (freekman) [inline tests/com/clouway/templates/ServletCounterTest.java]: where is the @Rule

File: tests/com/clouway/templates/ServletCounterTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+      will(returnValue(("https://github.com/clouway")));
+
+      oneOf(response).setContentType("text/html");
+
+      oneOf(response).getWriter();
+      will(returnValue(writer));
+    }});
+
+    counter.doGet(request, response);
+    String actual = counter.getHtml("web/WEB-INF/links.html",writer);
```
Comment (freekman) [inline tests/com/clouway/templates/ServletCounterTest.java]: format

File: src/com/clouway/adapter/server/jetty/Jetty.java (hunk 2 above)
Comment (Stanislava88) [inline src/com/clouway/adapter/server/jetty/Jetty.java]: fixed

File: src/com/clouway/links/ServletAccessCounter.java (hunk 3 above)
Comment (Stanislava88) [inline src/com/clouway/links/ServletAccessCounter.java]: fixed

File: src/com/clouway/links/ServletAccessCounter.java (hunk 4 above)
Comment (Stanislava88) [inline src/com/clouway/links/ServletAccessCounter.java]: renamed

File: src/com/clouway/links/ServletAccessCounter.java (hunk 6 above)
Comment (Stanislava88) [inline src/com/clouway/links/ServletAccessCounter.java]: done

File: src/com/clouway/links/ServletAccessCounter.java (hunk 7 above)
Comment (Stanislava88) [inline src/com/clouway/links/ServletAccessCounter.java]: done

File: tests/com/clouway/templates/ServletCounterTest.java (hunk 9 above)
Comment (Stanislava88) [inline tests/com/clouway/templates/ServletCounterTest.java]: fixed

File: pom.xml [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+    <!--Test-->
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.12</version>
+      <scope>test</scope>
+    </dependency>
+
+    <!--jMock-->
```
Comment (freekman) [inline pom.xml]: is jmock part of the tests and if so why do you have a separate comment for it

Comment (Stanislava88) [inline pom.xml]: removed

**Comment by freekman** [comment]
LGTM

File: test/com/clouway/templates/ServletCounterTest.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+    counter.doGet(request, response);
+
+    String actual = counter.getHtml("web/WEB-INF/links.html", writer);
+
+    assertThat(actual, containsString(" This link is accessed: 1 times"));
+  }
+
+  @Test
+  public void MultipleLinkVisits() throws Exception {
```
Comment (stefan-dimitrov) [inline test/com/clouway/templates/ServletCounterTest.java]: Method name starts with capital letter.

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        </configuration>
+      </plugin>
+
+    </plugins>
+  </build>
+
+  <dependencies>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
```
Comment (stefan-dimitrov) [inline pom.xml]: This dependency is already defined below.

**Comment by stefan-dimitrov** [comment]
LGTM.

File: test/com/clouway/templates/ServletCounterTest.java (hunk 12 above)
Comment (Stanislava88) [inline test/com/clouway/templates/ServletCounterTest.java]: fixed

File: pom.xml (hunk 13 above)
Comment (Stanislava88) [inline pom.xml]: removed

**Pull request merged**